### PR TITLE
from duplicate to main

### PR DESCRIPTION
**Pull request by amit-jesani**
delete update method and add class1 method

### CODE REVIEW

**Comment by amit-jesani** [comment]
### Review for Changes in `dy_new_plugin/dy_new_plugin/BusinessLogic/Class1.cs`:

```diff
@@ -0,0 +1,8 @@
+﻿using System;
+
+public class Class1
+{
+	public Class1()
+	{
+	}
+}
```

The provided code snippet defines a basic C# class named `Class1` within the namespace `dy_new_plugin.dy_new_plugin.BusinessLogic`. Here's a concise review:

1. **Class Name**: The name `Class1` is very generic. Consider renaming the class to something more descriptive related to its purpose or functionality.

2. **Empty Constructor**: The constructor is empty and does not perform any initialization. If no specific construction logic is needed, you can omit the constructor entirely as C# provides a default parameterless constructor if none is defined.

3. **Namespace**: The class is not explicitly defined within a namespace, which can lead to conflicts if integrated into larger projects. Consider wrapping the class in a namespace that matches its location or purpose, e.g., `namespace dy_new_plugin.dy_new_plugin.BusinessLogic`.

4. **Access Modifiers**: The class is declared with the default access modifier, which is `internal` in C#. If you intend for this class to be accessible from other assemblies, you should explicitly declare it as `public`.

5. **File Structure**: The file structure and naming convention (`Class1.cs`) should match the class name for clarity and maintainability.

6. **Using Directives**: The `using System;` directive is included but not utilized in the provided code snippet. If the `System` namespace is not used in the subsequent development, consider removing this directive to clean up the code.

7. **Documentation**: There are no comments or XML documentation comments. Adding comments or documentation would be beneficial for maintainability and clarity, especially if the class is part of a larger project or library.

8. **Optimizations and Improvements**: Since the class currently does not contain any properties, methods, or fields, it's hard to suggest specific optimizations. However, consider future-proofing the class by thinking about potential features and designing accordingly.

In summary, while the code is syntactically correct, there are several improvements related to naming, structure, and design practices that could enhance the clarity, maintainability, and functionality of the class.

**Comment by amit-jesani** [comment]
### Review for Changes in `dy_new_plugin/dy_new_plugin/BusinessLogic/Class1.cs`:

```diff
@@ -0,0 +1,29 @@
+﻿using System;
+
+public class Class1
+{
+
+
+        public void deleteaccount()
+        {
+            if 
+            (_target.LogicalName == "account" && _context.MessageName == "Delete")
+            {
+                    try
+                {
+                  EntityReference contactEntityRef = _preImage.GetValue<EntityReference>("primarycontactid", null);
+                    if (contactEntityRef != null)
+                 {
+                        _context.Service.Delete(contactEntityRef.LogicalName, contactEntityRef.Id);
+                  }
+                }
+
+                catch (Exception ex)
+                {
+                    throw new InvalidPluginExecutionException(ex.Message);
+                }
+            }
+
+        }
+    
+}
```

The provided code snippet from the file "dy_new_plugin/dy_new_plugin/BusinessLogic/Class1.cs" contains several issues and areas for improvement regarding syntax, style, and potential bugs. Here's a concise review:

1. **Code Formatting and Style:**
   - The indentation and spacing are inconsistent. Proper indentation and spacing improve readability and maintainability.
   - Method name `deleteaccount` should follow PascalCase naming conventions, i.e., `DeleteAccount`.

2. **Error Handling:**
   - Catching a general `Exception` and rethrowing a new exception with the original message loses the stack trace. Consider logging the full exception or passing the original exception as an inner exception.

3. **Missing Class Members:**
   - The code uses `_target`, `_context`, and `_preImage` but none of these members are defined in the provided class. Ensure these are either passed to the method as parameters or defined at the class level.

4. **Method Accessibility:**
   - The method `deleteaccount` is public but does not specify any class or interface that it overrides or implements. If this method is intended to be an override or part of an interface, it should be explicitly stated.

5. **Logical Checks:**
   - The condition checks `_target.LogicalName == "account"` and `_context.MessageName == "Delete"`. Ensure that `_target` and `_context` are not null before accessing their properties to avoid `NullReferenceException`.

6. **Use of Magic Strings:**
   - The strings "account", "Delete", and "primarycontactid" are used directly. Consider defining them as constants or retrieving them from a configuration setting to avoid errors due to typos and to make maintenance easier.

7. **Exception Handling Specificity:**
   - The exception handling is very generic. Depending on what `_context.Service.Delete` can throw, it might be beneficial to catch more specific exceptions to handle different failure scenarios differently.

8. **Code Comments and Documentation:**
   - There are no comments or documentation. Adding summary comments to the class and method would improve the understandability of the code, especially explaining the purpose and the logic behind the deletion criteria.

9. **Potential Bug with Exception Handling:**
   - Rethrowing the exception with `new InvalidPluginExecutionException(ex.Message)` might not include all relevant information about the original exception. Consider using `new InvalidPluginExecutionException("Custom message", ex)` to

**Comment by amit-jesani** [comment]
### Review for Changes in `dy_new_plugin/dy_new_plugin/BusinessLogic/createInCRM.cs`:

```diff
@@ -1,4 +1,5 @@
 ﻿using Microsoft.Xrm.Sdk;
+using Microsoft.Xrm.Sdk.Query;
 using System;
 
 namespace dy_new_plugin.BusinessLogic
@@ -60,29 +61,33 @@ public void updateDEmo()
             {
                 try
                 {
-                    EntityReference DemoEntityRef = _preImage.GetValue<EntityReference>("dt_demo_2id", null);
+                    EntityReference DemoEntityRef = _target.GetValue<EntityReference>("dt_demo_2id", null);
                     if (DemoEntityRef == null)
                     {
                         throw new InvalidPluginExecutionException("DemoEntityRef is null");
                     }
                     _context.Trace("DemoEntityRef: " + DemoEntityRef.Id);
 
-                    Entity deMOEntity = _context.Retrieve(DemoEntityRef, new string[] { "dt_demo_name" });
+                    Entity deMOEntity = _context.Service.Retrieve("dt_demo_2", DemoEntityRef.Id, new ColumnSet("dt_demo_name", "dt_demo_refernace"));
                     if (deMOEntity == null)
                     {
                         throw new InvalidPluginExecutionException("deMOEntity is null");
                     }
 
-                    deMOEntity["dt_demo_name"] = (_target.Contains("dt_demo_name") ? _target["dt_demo_name"] : _preImage["dt_demo_name"]);
-                    _context.Trace("DemoEntityRef: " + deMOEntity["dt_demo_name"]);
+                    deMOEntity.SetValue("dt_demo_name", _target["dt_demo_name"]);
+                    deMOEntity.SetValue("dt_demo_refernace", _target["dt_testparentrefernace"]);
+                    _context.Service.Update(deMOEntity);
+
+
+                    //deMOEntity["dt_demo_name"] = (_target.Contains("dt_demo_name") ? _target["dt_demo_name"] : _preImage["dt_demo_name"]);
+                    //_context.Trace("DemoEntityRef: " + deMOEntity["dt_demo_name"]);
 
                     //deMOEntity["dt_demo_refernace"] = (_target.Contains("dt_testparentrefernace") ? _target["dt_testparentrefernace"] : _preImage["dt_testparentrefernace"]);
 
                     // update lookup field using preimage
                     //Guid id = _target.GetAttributeValue<EntityReference>("dt_testparentrefernace").Id;
                     //deMOEntity["dt_demo_refernace"] = new EntityReference("dt_testparentrefernace", id);
 
-                    _context.Service.Update(deMOEntity);
                 }
                 catch (Exception ex)
                 {
```

Here are some observations and suggestions for the code snippet provided:

1. **Namespace Import**:
   - The addition of `using Microsoft.Xrm.Sdk.Query;` is appropriate for using `ColumnSet` in the `Retrieve` method.

2. **Entity Retrieval**:
   - Changing `_preImage.GetValue<EntityReference>("dt_demo_2id", null)` to `_target.GetValue<EntityReference>("dt_demo_2id", null)` alters the source of the entity reference. Ensure this change aligns with the intended logic, as `_target` and `_preImage` might contain different data depending on the plugin stage and whether the field was modified in the current operation.

3. **Error Handling**:
   - Throwing an exception if `DemoEntityRef` or `deMOEntity` is null is a good practice for early exit and avoiding null reference errors later in the code.

4. **Entity Update**:
   - The method call `deMOEntity.SetValue("dt_demo_name", _target["dt_demo_name"]);` suggests a custom extension method or a modified API usage. Normally, you would directly set the attribute like `deMOEntity["dt_demo_name"] = _target["dt_demo_name"];`. If `SetValue` is not a defined method, revert to the standard way.
   - The field `"dt_demo_refernace"` seems to be a typo. Confirm if it should be `"dt_demo_reference"`. Consistency in field names is crucial to avoid runtime errors.
   - The use of `ColumnSet` in the `Retrieve` method now includes `"dt_demo_refernace"`, which is good if you need this field for logic not shown here.

5. **Commented Code**:
   - The commented-out lines for setting `dt_demo_name` and `dt_demo_refernace` using ternary operations suggest an alternative approach to handle cases where `_target` might not contain the new values, and fallback to `_preImage` is needed. If this logic is required (handling cases where `_target` does not have these fields), it should be reinstated or handled appropriately.

6. **Optimization**:
   - Directly updating the entity after setting the values is efficient. However, ensure that all necessary fields are correctly set before calling `Update`.
   - The trace statements are commented out. If debugging or logging is required, consider enabling these.

7. **Syntax and Typos**:
   - Ensure that

**Comment by amit-jesani** [comment]
### Additional Context for `dy_new_plugin/dy_new_plugin/BusinessLogic/createInCRM.cs`:

### Code Review Report

#### File: `dy_new_plugin/dy_new_plugin/BusinessLogic/createInCRM.cs`

#### Overview
The file contains a class `createInCRM` that appears to handle CRM operations related to the creation and update of entities within a Dynamics CRM environment. The class methods focus on creating and updating entities based on the context provided by a plugin execution.

#### Specific Review Points

1. **Class and Method Naming Conventions**
   - The class name `createInCRM` should follow PascalCase naming convention and ideally be more descriptive, e.g., `CrmEntityCreator`.
   - Method names like `createdeMoFAbS` and `updateDEmo` are not clear and do not follow standard naming conventions. They should be renamed to more descriptive and conventional names like `CreateDemoEntity` and `UpdateDemoEntity`.

2. **Code Duplication and Redundancy**
   - The check for `_context.MessageName == "Create"` is redundant in `createdeMoFAbS` method as it's already checked in the outer `if` condition.

3. **Error Handling**
   - The exception handling only rethrows a new exception with the original message. It might be more useful to include additional context-specific information in the error message to aid in debugging.
   - In `updateDEmo`, the exception message `"This plugin is only for account entity and contact can't be created"` is misleading since the method deals with `dt_demo_fab` entities.

4. **Performance Considerations**
   - In `updateDEmo`, the method retrieves and updates an entity but does not conditionally check fields before updating. This could lead to unnecessary update calls even if the data has not changed.
   - Use of `EntityReference` and checks like `if (_target.Contains("dt_demo_name"))` should be optimized to avoid unnecessary retrievals or operations.

5. **Code Clarity and Maintainability**
   - Inline comments are sparse and sometimes unclear, which can make maintenance difficult. More descriptive comments should be added to explain the purpose of code blocks and logic decisions.
   - The commented-out code should be removed if it's not intended to be used, as it clutters the codebase and can lead to confusion.

6. **Security and Best Practices**
   - Directly throwing `InvalidPluginExecutionException` with `ex.Message` could potentially expose sensitive information. Consider sanitizing or standardizing error messages.
   - Ensure that all user inputs or external data manipulations are validated to prevent issues like SQL injections or cross-site scripting in other layers.

#### Recommendations for Improvement

1. **Refactor Naming Conventions**: Update class and method names to be more descriptive and follow C# naming conventions.
2. **Remove Redundant Code**: Eliminate unnecessary checks and commented-out code to streamline the codebase.
3. **Enhance Error Handling**: Include more context in error messages and consider logging errors for further analysis.
4. **Optimize Performance**: Add checks to ensure that only necessary updates are performed.
5. **Improve Code Documentation**: Add meaningful comments and documentation to make the code easier to understand and maintain.

#### Conclusion
The file provides basic functionality for creating and updating CRM entities but needs improvements in terms of naming, performance optimization, error handling, and overall code quality to ensure maintainability and scalability.

**Comment by amit-jesani** [comment]
### Review for Changes in `dy_new_plugin/dy_new_plugin/BusinessLogic/Class1.cs`:

```diff
@@ -0,0 +1,29 @@
+﻿using System;
+
+public class Class1
+{
+
+
+        public void deleteaccount()
+        {
+            if 
+            (_target.LogicalName == "account" && _context.MessageName == "Delete")
+            {
+                    try
+                {
+                  EntityReference contactEntityRef = _preImage.GetValue<EntityReference>("primarycontactid", null);
+                    if (contactEntityRef != null)
+                 {
+                        _context.Service.Delete(contactEntityRef.LogicalName, contactEntityRef.Id);
+                  }
+                }
+
+                catch (Exception ex)
+                {
+                    throw new InvalidPluginExecutionException(ex.Message);
+                }
+            }
+
+        }
+    
+}
```

The provided code snippet for `Class1.cs` in the `dy_new_plugin` project has several issues and areas for improvement related to syntax, exception handling, and overall code quality. Below is a concise review:

1. **Code Formatting and Style:**
   - The indentation is inconsistent. Proper indentation should be maintained for better readability.
   - There are unnecessary blank lines (e.g., lines 4, 5, 27) that do not contribute to code clarity and should be removed.

2. **Method Naming:**
   - The method name `deleteaccount` should follow C# naming conventions, which recommend PascalCase for methods. It should be renamed to `DeleteAccount`.

3. **Error Handling:**
   - The exception handling in the method rethrows an exception with only the message of the original exception. This can lead to loss of stack trace information. Consider rethrowing the original exception or logging the complete exception details before throwing a new one.

4. **Variable and Member Access:**
   - The code snippet uses `_target`, `_context`, and `_preImage` without declaring them. Ensure these are either properly declared in the class or passed as parameters to the method.
   - The method assumes that `_target.LogicalName` and `_context.MessageName` are accessible and correctly set. This could lead to runtime errors if they are not properly initialized.

5. **Exception Specificity:**
   - The catch block captures all exceptions (`Exception ex`) which is generally not recommended. It is better to catch more specific exceptions related to the operations being performed (e.g., `FaultException` for service calls).

6. **Use of Null Coalescing Operator:**
   - The method `GetValue` is called with a second parameter `null` which seems intended to be a default value. If `_preImage.GetValue` is meant to handle nulls, consider using the null coalescing operator (`??`) to provide default values more explicitly.

7. **Code Comments and Documentation:**
   - The method lacks comments or documentation. Adding summary comments explaining the purpose of the method and describing its parameters and exceptions can improve maintainability and understandability.

8. **Potential Logical Errors:**
   - The method checks if the logical name is "account" and the message name is "Delete", but does not handle cases where these conditions are not met. It might be useful to include an `else` clause or other error handling to manage

**Comment by amit-jesani** [comment]
### Review for Changes in `dy_new_plugin/dy_new_plugin/BusinessLogic/createInCRM.cs`:

```diff
@@ -1,4 +1,5 @@
 ﻿using Microsoft.Xrm.Sdk;
+using Microsoft.Xrm.Sdk.Query;
 using System;
 
 namespace dy_new_plugin.BusinessLogic
@@ -8,7 +9,7 @@ public class createInCRM
         private Entity _target;
         private PluginContext<Entity> _context;
         private Entity _preImage;
-        public createInCRM(PluginContext<Entity> context) 
+        public createInCRM(PluginContext<Entity> context)
         {
             _context = context;
             _target = context.Target;
@@ -26,17 +27,14 @@ public void createdeMoFAbS()
                     {
                         Entity demoentiti = new Entity("dt_demo_2");
 
-
                         demoentiti.SetValue("dt_demo_name", _target["dt_demo_name"]);
-                        //demoentiti["dt_demo_refernace"] = _target["dt_testparentrefernace"];
-
-                        // set lookup field
-                                Guid id = _target.GetValue<EntityReference>("dt_testparentrefernace", null).Id;
-                                    demoentiti.SetValue("dt_demo_refernace", new EntityReference("cr4fb_test", id));
 
+                        Guid id = _target.GetValue<EntityReference>("dt_testparentrefernace", null).Id;
+                        demoentiti.SetValue("dt_demo_refernace", new EntityReference("cr4fb_test", id));
 
+                        demoentiti.SetValue("dt_demofabreference", new EntityReference("dt_demo_fab", _target.Id);
 
-                                Guid DEMOid = _context.Service.Create(demoentiti);
+                        Guid DEMOid = _context.Service.Create(demoentiti);
 
                         //if (contactId != Guid.Empty)
                         //{
@@ -48,41 +46,47 @@ public void createdeMoFAbS()
                     {
                         throw new InvalidPluginExecutionException(ex.Message);
                     }
-             }
+                }
             }
-                 }
+        }
 
         // update demo entity when the demo_fabs is updated
-        
+
         public void updateDEmo()
         {
             if (_target.LogicalName == "dt_demo_fab" && _context.MessageName == "Update")
             {
                 try
                 {
-                    EntityReference DemoEntityRef = _preImage.GetValue<EntityReference>("dt_demo_2id", null);
+                    EntityReference DemoEntityRef = _target.GetAttributeValue<EntityReference>("dt_demo_2id");
                     if (DemoEntityRef == null)
                     {
                         throw new InvalidPluginExecutionException("DemoEntityRef is null");
                     }
                     _context.Trace("DemoEntityRef: " + DemoEntityRef.Id);
 
-                    Entity deMOEntity = _context.Retrieve(DemoEntityRef, new string[] { "dt_demo_name" });
+                    Entity deMOEntity = _context.Service.Retrieve("dt_demo_2", DemoEntityRef.Id, new ColumnSet("dt_demo_name"));
                     if (deMOEntity == null)
                     {
                         throw new InvalidPluginExecutionException("deMOEntity is null");
                     }
 
-                    deMOEntity["dt_demo_name"] = (_target.Contains("dt_demo_name") ? _target["dt_demo_name"] : _preImage["dt_demo_name"]);
-                    _context.Trace("DemoEntityRef: " + deMOEntity["dt_demo_name"]);
+                    if (_target.Contains("dt_demo_name"))
+                    {
+                        deMOEntity["dt_demo_name"] = _target["dt_demo_name"];
+                    }
+                    _context.Service.Update(deMOEntity);
+
+
+                    //deMOEntity["dt_demo_name"] = (_target.Contains("dt_demo_name") ? _target["dt_demo_name"] : _preImage["dt_demo_name"]);
+                    //_context.Trace("DemoEntityRef: " + deMOEntity["dt_demo_name"]);
 
                     //deMOEntity["dt_demo_refernace"] = (_target.Contains("dt_testparentrefernace") ? _target["dt_testparentrefernace"] : _preImage["dt_testparentrefernace"]);
 
                     // update lookup field using preimage
                     //Guid id = _target.GetAttributeValue<EntityReference>("dt_testparentrefernace").Id;
                     //deMOEntity["dt_demo_refernace"] = new EntityReference("dt_testparentrefernace", id);
 
-                    _context.Service.Update(deMOEntity);
                 }
                 catch (Exception ex)
                 {
@@ -95,7 +99,6 @@ public void updateDEmo()
             }
         }
 
-
         //public void deleteaccount()
         //{
         //    if (_target.LogicalName == "account" && _context.MessageName == "Delete")
```

Here's a concise review of the changes in the file `createInCRM.cs`:

1. **Syntax and Typographical Errors:**
   - Missing closing parenthesis in the line `demoentiti.SetValue("dt_demofabreference", new EntityReference("dt_demo_fab", _target.Id);`. It should be `demoentiti.SetValue("dt_demofabreference", new EntityReference("dt_demo_fab", _target.Id));`.

2. **Code Optimization and Best Practices:**
   - The use of `GetValue` method is inconsistent. It's better to use a consistent method for accessing entity attributes, either `GetValue` or direct indexing like `_target["dt_demo_name"]`.
   - Commented out code should be removed if it's no longer needed to keep the codebase clean and maintainable.
   - The exception handling in both methods `createdeMoFAbS` and `updateDEmo` only throws the exception message. It might be more useful to include more context or use the original exception as an inner exception to preserve the stack trace.
   - In the `updateDEmo` method, the retrieval of `DemoEntityRef` uses `_target.GetAttributeValue<EntityReference>("dt_demo_2id")` which is more concise and preferable over the previous `_preImage.GetValue<EntityReference>`.
   - The update logic in `updateDEmo` is improved by checking if `_target` contains "dt_demo_name" before attempting to update it, which is a good practice to avoid unnecessary updates.

3. **Potential Bugs and Issues:**
   - In the `createdeMoFAbS` method, the line `demoentiti.SetValue("dt_demo_refernace", new EntityReference("cr4fb_test", id));` assumes that `id` will not be null. If `_target.GetValue<EntityReference>("dt_testparentrefernace", null)` returns null, this will cause a null reference exception when trying to access `.Id`.
   - The logic in `updateDEmo` does not handle the scenario where `_target` might not contain "dt_demo_name" and `_preImage` is not used to provide a fallback as the commented code suggests. This could lead to data inconsistency if "dt_demo_name" is supposed to be maintained.

4. **General Observations:**
   - The class and method names (`createInCRM`, `createdeMoFAbS`, `updateDEmo`) do

**Comment by amit-jesani** [comment]
### Additional Context for `dy_new_plugin/dy_new_plugin/BusinessLogic/createInCRM.cs`:

### Code Review and Analysis

#### General Observations
- The class `createInCRM` contains methods to handle creation and updating of CRM entities based on plugin context. The naming conventions and structuring of the class and methods could be improved for better readability and maintainability.

#### Specific Observations and Recommendations

**1. Class and Method Naming**
   - The class name `createInCRM` should follow PascalCase naming conventions for C# classes, e.g., `CreateInCrm`.
   - Method names like `createdeMoFAbS` and `updateDEmo` should be more descriptive and follow CamelCase, e.g., `CreateDemoEntity` and `UpdateDemoEntity`.

**2. Code Duplication and Redundancy**
   - The check for `_context.MessageName == "Create"` is redundant in `createdeMoFAbS` as it's already inside a conditional block checking the same.
   - Consider extracting repeated logic into private helper methods to reduce duplication and improve maintainability.

**3. Error Handling and Logging**
   - The exception handling currently just rethrows a new exception with the original message. It would be beneficial to include more context-specific information in the error messages.
   - Utilize `_context.Trace` for logging important steps or variable values to aid in debugging.

**4. Use of Magic Strings**
   - There are multiple instances where entity names and field names are hardcoded (e.g., `"dt_demo_fab"`, `"dt_demo_name"`). Consider defining these as constants or fetching them from a configuration source to avoid errors and ease future changes.

**5. Commented-Out Code**
   - There are several commented-out lines of code. If these are not required, they should be removed to clean up the codebase. If they might be needed for future reference, consider adding comments to explain why they are retained.

**6. Performance Considerations**
   - In `updateDEmo`, the method retrieves an entity and updates it. If only certain fields are subject to change, ensure that only those necessary fields are retrieved to optimize performance.
   - The method `updateDEmo` throws a generic exception if it's not handling an update for `dt_demo_fab`. This could be handled more gracefully, or the method could be designed to only be triggered for the appropriate entity type.

**7. Code Clarity and Readability**
   - Use meaningful variable names. For instance, `demoentiti` could be renamed to `demoEntity`.
   - The method `updateDEmo` uses a variable `DemoEntityRef` which could be renamed to `demoEntityReference` for clarity.

**8. Robustness**
   - In `updateDEmo`, before updating the entity, check if the target contains the fields that are being set to avoid potential null reference exceptions.

### Optimized Example for `createdeMoFAbS` Method

```csharp
public void CreateDemoEntity()
{
    if (_target.LogicalName == "dt_demo_fab" && _context.MessageName == "Create")
    {
        try
        {
            Entity demoEntity = new Entity("dt_demo_2");
            demoEntity["dt_demo_name"] = _target.GetAttributeValue<string>("dt_demo_name");

            Guid parentId = _target.GetAttributeValue<EntityReference>("dt_testparentrefernace").Id;
            demoEntity["dt_demo_refernace"] = new EntityReference("cr4fb_test", parentId);

            Guid demoId = _context.Service.Create(demoEntity);
        }
        catch (Exception ex)
        {
            _context.Trace($"Error creating demo entity: {ex.Message}");
            throw new InvalidPluginExecutionException($"Failed to create demo entity: {ex.Message}");
        }
    }
}
```

This refactoring includes improved naming, reduced redundancy, and enhanced error handling with contextual logging.
